spec/controllers/packages_controller_spec.rb

```diff
@@ -1,51 +1,180 @@
+      state: "FL",
+      city: "Ft Lauderdale",
+      zip: "33316"
+    },
+    package: {
+      weight: 100,
+      length: 10,
+      width: 10,
+      height: 10,
+      units: "metric"
+    }
+  }
+end
+
+  let(:bad_destination_params) do {
+    destination: {
+      country: "",
+      state: "WA",
+      city: "Seattle",
+      zip: "98112"
+    },
+    origin: {
+      country: "US",
+      state: "FL",
+      city: "Ft Lauderdale",
+      zip: "33316"
+    },
+    package: {
+      weight: 100,
+      length: 10,
+      width: 10,
+      height: 10,
+      units: "metric"
+    }
+  }
+  end
+
+  let(:false_params) do {
+    destination: {
+      country: "US",
+      state: "FL",
+      city: "Ft Lauderdale",
+      zip: "33316"
+    },
+    origin: {
+      country: "US",
+      state: "FL",
+      city: "Ft Lauderdale",
+      zip: "33316"
+    },
+    package: {
+      weight: 100,
+      length: 10,
+      width: 10,
+      height: 10,
+      units: "metric"
+    }
+  }
+end
+
+  let(:keys) { ["ups", "usps"]}
+
+   describe "GET 'rates'" do
     it "is successful" do
-      get :rates
+      get :rates, params
       expect(response.response_code).to eq 200
     end
 
     it "returns json" do
-      get :rates
+      get :rates, params
       expect(response.header['Content-Type']).to include 'application/json'
     end
 
     context "the returned json object" do
       before :each do
-        get :rates
+        get :rates, params
         @response = JSON.parse response.body
       end
 
       it "has the right keys" do
         expect(@response.keys.sort).to eq keys
       end
-
-      it "has all of Rosa's info" do
-        keys.each do |key|
-          expect(@response[key]).to eq rosa[key]
-        end
-      end
     end
 
-    context "no rates found" do
-      before :each do
-        get :rates
+    context "bad params passed" do
+      it "returns an error if incorrect package params are passed" do
+        get :rates, bad_package_params
+        @response = JSON.parse response.body
+        expect(@response).to eq ["Incorrect or missing parameters for package"]
       end
 
-      it "is successful" do
-        expect(response).to be_successful
+      it "returns an error if incorrect origin params are passed" do
+        get :rates, bad_origin_params
+        @response = JSON.parse response.body
+        expect(@response).to eq ["Incorrect or missing parameters for origin address"]
       end
 
-      it "returns a 204 (no content)" do
-        expect(response.response_code).to eq 204
+      it "returns an error if incorrect package params are passed" do
+        get :rates, bad_destination_params
+        @response = JSON.parse response.body
+        expect(@response).to eq ["Incorrect or missing parameters for destination address"]
       end
 
-      it "expects the response body to be an empty array" do
-        expect(response.body).to eq "[]"
-      end
+    end
+
+    context "no rates found" do
```

Comment:
It would have been interesting to see what this would yield if ActiveShipping was unable to provide any options and how that would get communicated to the user (like if a package was millions of lbs).